numpy/lib/recfunctions.py

```diff
@@ -972,6 +1022,39 @@ def structured_to_unstructured(arr, dtype=None, copy=False, casting='unsafe'):
                                  'itemsize': arr.dtype.itemsize})
     arr = arr.view(flattened_fields)
 
+    # we only allow a few types to be unstructured by manipulating the
+    # strides, because we know it won't work with, for example, np.matrix nor
+    # np.ma.MaskedArray.
+    can_view = type(arr) in (np.ndarray, np.recarray, np.memmap)
+    if (not copy) and can_view and all(dt.base == out_dtype for dt in dts):
+        # all elements have the right dtype already; if they have a common
+        # stride, we can just return a view
+        common_stride = _common_stride(offsets, counts, out_dtype.itemsize)
+        if common_stride is not None:
+            wrap = arr.__array_wrap__
+
+            new_shape = arr.shape + (sum(counts), out_dtype.itemsize)
+            new_strides = arr.strides + (abs(common_stride), 1)
+
+            arr = arr[..., np.newaxis].view(np.uint8)  # view as bytes
+            arr = arr[..., min(offsets):]  # remove the leading unused data
+            arr = np.lib.stride_tricks.as_strided(arr,
+                                                  new_shape,
+                                                  new_strides,
+                                                  subok=True)
+
+            # cast and drop the last dimension again
+            arr = arr.view(out_dtype)[..., 0]
```

Comment:
Just noticed something else: Is there a reason I'm missing for not using `np.ndarray` instead of the last few lines:

```python
new_shape = arr.shape + (sum(counts),)
new_strides = arr.strides + (abs(common_stride),)
arr = np.ndarray(buffer=arr, dtype=out_dtype, shape=new_shape, strides=new_strides, offset=min(offsets)).view(type(a))
```

Reply:
I was not aware that constructor existed. I will try that out.

Reply:
I'm facing two issues with that:

1. The `np.ndarray()` throws "ValueError: ndarray is not contiguous" in some cases. At least it does when the input array is reversed; maybe in other cases too.
2. I couldn't find a way to preserve the `memmap` subclass yet. Just using `view()` to get it doesn't seem very useful, because that would lose its attributes like `filename` (if it even works). And its `__array_wrap__()` seems to deliberately return a plain ndarray instead.

Reply:
Hmm, thanks for checking. Then the code is fine as-is.

Somehow it seems like an oversight in the API that the `np.ndarray(buffer=xxx)` can't handle non-contiguous inputs. The memmap issue could probably be fixed by appropriate call to `__new__`, but that's a moot point.